File: ruby/single_number.rb
# https://leetcode.com/problems/single-number/

# need to return num that is either different from previous, next numbers
# or the same as previous, next numbers
# as long as it isn't a number that appears twice

def single_number(nums)
    nums.sort!
    nums.each_with_index do |num, index|
        
        # when number is first (index == 0)
        if index == 0
            next_num = nums[index + 1]
            return num if num != next_num
            
        # when number is in the middle
        else
            prev_num = nums[index - 1]
            next_num = nums[index + 1]
            return num if (num == prev_num) && (num == next_num)
            return num if (num != prev_num) && (num != next_num)
        end
    end
end

File: go/single_number.go
// https://leetcode.com/problems/single-number/

package main

import (
	"fmt"
	"sort"
)

func singleNumber(nums []int) int {
	sort.Ints(nums)
	answer := 0

	for i := 0; i < len(nums); i++ {

		if i == len(nums)-1 {
			answer = nums[i]
			break
		} else if i == 0 {
			if nums[i] != nums[i+1] {
				answer = nums[i]
				break
			}
		} else {
			if (nums[i] == nums[i-1]) && (nums[i] == nums[i+1]) {
				answer = nums[i]
				break
			}

			if (nums[i] != nums[i-1]) && (nums[i] != nums[i+1]) {
				answer = nums[i]
				break
			}
		}
	}
	return answer
}

